### DISCUSSION TITLE
Solving convex quadratic programs with PDLP

### DISCUSSION THREAD

**Discussion by stephane-caron** [google/or-tools#4026] [Linear Solver questions]
**What language and solver does this apply to?**

C++, Python / PDLP

**Describe the problem you are trying to solve.**

Solve a convex quadratic program, as described in [Mathematical background for PDLP](https://developers.google.com/optimization/lp/pdlp_math).

The objective matrix of a QP can be any symmetric positive semidefinite matrix, but presently the objective matrix in PDLP must be diagonal and non-negative:

https://github.com/google/or-tools/blob/8768ed7a43f8899848effb71295a790f3ecbe2f2/ortools/pdlp/quadratic_program.h#L129

This is accurately documented in the code, but there is then a mismatch with the level of generality described in the mathematical background note.

https://github.com/google/or-tools/blob/8768ed7a43f8899848effb71295a790f3ecbe2f2/ortools/pdlp/quadratic_program.h#L41

**Describe the solution you'd like**

The ability to set a general (symmetric positive semidefinite matrix) cost matrix.

**Additional context**

Adding a solver interface for PDLP in https://github.com/qpsolvers/qpsolvers/pull/260

**Comment by dapplegate** [comment]
Thank you for pointing this out. The documentation was out of date; the code was correct. PDLP does not handle general symmetric positive semidefinite matrices, only diagonal ones. We're updating the documentation to correct this. It is unlikely that PDLP will support general PSD $Q$ in the near future.

If you want to use PDLP for a general symmetric positive semidefinite matrix, and are able to compute a factorization (for example, Cholesky) of the matrix (Q = R^T R), then you can convert your problem into one with a diagonal objective matrix by introducing additional variables $z$ defined by constraints $R x - z = 0$, so that $x^T S x = z^T z$. This doesn't necessarily provide a good solution for using PDLP as a general-purpose QP solver, since the factorization may be expensive to compute, and $R$ may have many more nonzeros than $Q$ making the iterations much more expensive.

If `qpsolvers` can interface with Julia, then two alternatives for general QPs are https://github.com/google-research/FirstOrderLp.jl (the Julia prototype of PDLP) and https://github.com/jinwen-yang/PDQP (an implementation of https://arxiv.org/pdf/2311.07710.pdf).
